File: scripts/adapter/bridge/agentapi.py
import os
import json
import shutil
import subprocess


def get_binary():
    cmd = shutil.which("agentapi")
    if cmd:
        return cmd
    fallback = os.path.expanduser("~/.gemini/antigravity/bin/agentapi")
    if os.path.exists(fallback):
        return fallback
    return "agentapi"


def _call(action, *args, timeout=30, env=None):
    cmd = [get_binary(), action] + list(args)
    if env is None:
        env = os.environ.copy()
    return subprocess.check_output(cmd, env=env, stderr=subprocess.STDOUT, timeout=timeout)


def get_metadata(conv_id, timeout=10):
    result = _call("get-conversation-metadata", conv_id, timeout=timeout)
    data = json.loads(result.decode("utf-8"))
    return data.get("response", {}).get("conversationMetadata", {}).get("metadata", {})


def get_project_id(conv_id, default="11111111-1111-1111-1111-111111111111", timeout=10):
    try:
        meta = get_metadata(conv_id, timeout=timeout)
        return meta.get("projectId") or default
    except Exception:
        return default


def send_message(conv_id, prompt, timeout=120):
    env = os.environ.copy()
    env["ANTIGRAVITY_PROJECT_ID"] = "11111111-1111-1111-1111-111111111111"
    _call("send-message", conv_id, prompt, timeout=timeout, env=env)


def create_conversation(prompt, timeout=120):
    env = os.environ.copy()
    env["ANTIGRAVITY_PROJECT_ID"] = "11111111-1111-1111-1111-111111111111"
    result = _call("new-conversation", prompt, timeout=timeout, env=env)
    return json.loads(result.decode("utf-8"))

File: packages/adapter-antigravity/src/bridge/agentapi.ts
import * as fs from "node:fs";
import * as path from "node:path";
import { getAntigravityDir } from "./paths";
import { execSync, execFileSync } from "node:child_process";

export function getBinary(): string {
  try {
    const cmd = execSync("which agentapi").toString().trim();
    if (cmd) {
      return cmd;
    }
  } catch {
    // which failed, continue to fallback
  }
  const fallback = path.join(getAntigravityDir(), "bin", "agentapi");
  if (fs.existsSync(fallback)) {
    return fallback;
  }
  return "agentapi";
}

export function _call(action: string, args: string[], timeout: number = 30, env?: Record<string, string>): string {
  const cmd = [getBinary(), action].concat(args);
  const effectiveEnv = env ?? { ...process.env };
  const result = execFileSync(cmd[0], cmd.slice(1), {
    env: effectiveEnv,
    timeout: timeout * 1000,
  });
  return result.toString();
}

export function getMetadata(convId: string, timeout: number = 10): Record<string, unknown> {
  const result = _call("get-conversation-metadata", [convId], timeout);
  const data = JSON.parse(result);
  return data?.response?.conversationMetadata?.metadata ?? {};
}

export function getProjectId(
  convId: string,
  defaultVal: string = "11111111-1111-1111-1111-111111111111",
  timeout: number = 10
): string {
  try {
    const meta = getMetadata(convId, timeout);
    return (meta as any).projectId || defaultVal;
  } catch {
    return defaultVal;
  }
}

export function sendMessage(convId: string, prompt: string, timeout: number = 120): void {
  const env = { ...process.env, ANTIGRAVITY_PROJECT_ID: "11111111-1111-1111-1111-111111111111" };
  _call("send-message", [convId, prompt], timeout, env);
}

export function createConversation(prompt: string, timeout: number = 120, model?: string | null): unknown {
  const env = { ...process.env, ANTIGRAVITY_PROJECT_ID: "11111111-1111-1111-1111-111111111111" };
  const binary = getBinary();
  const cmdArgs = [binary, "new-conversation"];
  if (model) {
    cmdArgs.push(`--model=${model}`);
  }
  cmdArgs.push(prompt);
  const result = execFileSync(cmdArgs[0], cmdArgs.slice(1), {
    env,
    timeout: timeout * 1000,
  });
  return JSON.parse(result.toString());
}
